lib/trino-orc/src/main/java/io/trino/orc/metadata/statistics/StringStatisticsBuilder.java

```diff
@@ -158,16 +176,80 @@ public ColumnStatistics buildColumnStatistics()
     {
-        if (minOrMax == null || minOrMax.length() > stringStatisticsLimitInBytes) {
+        if (minOrMax == null || (!shouldCompactMinMax && minOrMax.length() > stringStatisticsLimitInBytes)) {
             return null;
         }
+        if (minOrMax.length() > stringStatisticsLimitInBytes) {
+            if (isMin) {
+                return StringCompactor.truncateMin(minOrMax, stringStatisticsLimitInBytes);
+            }
+            else {
+                return StringCompactor.computeMax(minOrMax, stringStatisticsLimitInBytes);
+            }
+        }
 
         // Do not hold the entire slice where the actual stats could be small
         if (minOrMax.isCompact()) {
             return minOrMax;
         }
         return Slices.copyOf(minOrMax);
     }
+
+    static final class StringCompactor
+    {
+        private StringCompactor() {}
+
+        public static Slice truncateMin(Slice slice, int maxBytes)
+        {
+            checkArgument(slice.length() > maxBytes);
+            int lastIndex = findLastCharacterInRange(slice, maxBytes);
+            return slice.slice(0, lastIndex);
+        }
+
+        public static Slice computeMax(Slice slice, int maxBytes)
+        {
+            if (maxBytes >= slice.length()) {
+                return slice;
+            }
+            int lastIndex = findLastCharacterInRange(slice, maxBytes);
+            int penultimateIndex = findLastCharacterInRange(slice, lastIndex - 1);
+            int lastCharCodePoint = getCodePointAt(slice, penultimateIndex);
+            while (lastCharCodePoint == MAX_CODE_POINT && penultimateIndex > 0) {
+                penultimateIndex = findLastCharacterInRange(slice, penultimateIndex - 1);
+                lastCharCodePoint = getCodePointAt(slice, penultimateIndex);
+            }
+
+            // whole string is made of MAX_CODE_POINT characters, we can't do anything here
+            if (penultimateIndex == 0 && lastCharCodePoint == MAX_CODE_POINT) {
+                return null;
+            }
+
+            lastCharCodePoint++;
+            Slice sliceToAppend = codePointToUtf8(lastCharCodePoint);
+            byte[] result = new byte[penultimateIndex + sliceToAppend.length()];
+            arraycopy(slice.byteArray(), 0, result, 0, penultimateIndex);
+            arraycopy(sliceToAppend.byteArray(), 0, result, penultimateIndex, sliceToAppend.length());
+            return Slices.wrappedBuffer(result);
+        }
+
+        private static int findLastCharacterInRange(Slice slice, int to)
+        {
+            byte[] bytes = slice.getBytes();
+            int pos = to;
+            while (pos >= 0) {
+                if (isUtfBlockStartChar(bytes[pos])) {
+                    return pos;
+                }
+                pos--;
+            }
+            throw new IllegalArgumentException("Provided byte array is not a valid utf8 string");
+        }
+
+        private static boolean isUtfBlockStartChar(byte b)
```

Comment:
`! io.airlift.slice.SliceUtf8#isContinuationByte` instead

Reply:
this is a private method

Reply:
right. let's copy it and mark with a comment it's copied from there

Reply:
isn't it cleaner to have a method that checks the condition we want instead of having one which result has to be inverted/negated?

Reply:
i wanted `isContinuationByte`  because it's a somewhat familar  thing for our codebase (even if it's a private method in arilift)

Reply:
ok, will do